### PR TITLE
Add appveyor setup to build and unit test

### PR DESCRIPTION
**Pull request by vdemeester**
The next step is to run e2e tests on windows too.

Adds a `make.ps1` powershell script to make it easy to compile and test.

```
.\scripts\make.ps1 -Binary
INFO: make.ps1 starting at 03/01/2018 14:37:28
INFO: Building...

 ________   ____  __.
 \_____  \ |    |/ _|
 /   |   \|      <
 /    |    \    |  \
 \_______  /____|__ \
         \/        \/

INFO: make.ps1 ended at 03/01/2018 14:37:30

.\scripts\make.ps1 -TestUnit
```

Related to #457

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by vdemeester** [comment]
Now the main question, is how to make appveyor build this pull-request… 🤔

**Comment by dnephin** [comment]
I created https://ci.appveyor.com/project/docker/cli but I think someone may still need to add the github webhook

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  - set Path=c:\go\bin;c:\gopath\bin;C:\Program Files (x86)\Bazaar\;C:\Program Files\Mercurial\%Path%
+  - go version
+  - go env
+
+build: false
+deploy: false
+
+test_script:
+  - go build github.com/docker/cli/cmd/docker
+  - for /f "" %%G in ('go list github.com/docker/cli/...') do ( go test %%G & IF ERRORLEVEL == 1 EXIT 1)
```
Comment (dnephin) [inline appveyor.yml]: I think we can still use bash, ex: https://github.com/stedolan/jq/blob/master/appveyor.yml, https://github.com/containerd/containerd/blob/master/.appveyor.yml

**Comment by dnephin** [comment]
The hook is added, so this is running now!

**Comment by vdemeester** [comment]
I'll add commits from https://github.com/docker/cli/pull/906 here I think (to make the first run green :angel:)

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  - git config --global core.autocrlf input
+
+# Build
+
+install:
+  # Install the specific Go version.
+  - rmdir c:\go /s /q
+  - appveyor DownloadFile https://storage.googleapis.com/golang/go%GOVERSION%.windows-amd64.msi
+  - msiexec /i go%GOVERSION%.windows-amd64.msi /q
+  - set Path=c:\go\bin;c:\gopath\bin;C:\Program Files (x86)\Bazaar\;C:\Program Files\Mercurial\%Path%
```
Comment (dnephin) [inline appveyor.yml]: I don't think this line is necessary. At least no the Bazaar/Mercurial parts

File: appveyor.yml (hunk 1 above)
Comment (dnephin) [reply appveyor.yml]: another example: https://github.com/gotestyourself/gotestyourself/pull/69/files#diff-180360612c6b8c4ed830919bbb4dd459R22

**Comment by dnephin** [comment]
Sorry about conflicts from the gotestyourself/assert PR.

File: cli/command/trust/key_load_test.go [hunk 3]
```diff
@@ -109,6 +115,7 @@ var testKeys = map[string][]byte{
 }
 
 func TestLoadKeyFromPath(t *testing.T) {
+	skip.If(t, func() bool { return runtime.GOOS == "windows" })
```
Comment (dnephin) [inline cli/command/trust/key_load_test.go]: The `func() bool {}` can be removed here, and below.

File: cli/compose/loader/full-struct_test.go [hunk 4]
```diff
@@ -389,3 +391,313 @@ func volumes() map[string]types.VolumeConfig {
 		},
 	}
 }
+
+func fullExampleYAML(workingDir, homeDir string) string {
+	return fmt.Sprintf(string(`version: "3.6"
```
Comment (dnephin) [inline cli/compose/loader/full-struct_test.go]: The `string()` cast is unnecessary here

File: cli/registry/client/client_test.go [hunk 5]
```diff
@@ -0,0 +1 @@
+package client
```
Comment (dnephin) [inline cli/registry/client/client_test.go]: empty file can be removed

File: scripts/make.ps1 [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    Author:  @vdemeester
+
+    Summary: Windows native build script.
+
+             It does however provided the minimum necessary to support parts of local Windows
+             development and Windows to Windows CI.
+
+             Usage Examples (run from repo root):
+                "hack\make.ps1 -Client" to build docker.exe client 64-bit binary (remote repo)
+                "hack\make.ps1 -TestUnit" to run unit tests
```
Comment (dnephin) [inline scripts/make.ps1]: old comments, wrong path.

File: scripts/make.ps1 [hunk 7]
```diff
@@ -0,0 +1,197 @@
+<#
+.NOTES
+    Author:  @vdemeester
```
Comment (justincormack) [inline scripts/make.ps1]: files shouldnt have authors, all the maintainers are responsible

File: scripts/make.ps1 [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+$ErrorActionPreference = "Stop"
+$ProgressPreference = "SilentlyContinue"
+$pushed=$False  # To restore the directory if we have temporarily pushed to one.
+
+# Utility function to get the commit ID of the repository
+Function Get-GitCommit() {
+    if (-not (Test-Path ".\.git")) {
+        # If we don't have a .git directory, but we do have the environment
+        # variable DOCKER_GITCOMMIT set, that can override it.
+        if ($env:DOCKER_GITCOMMIT.Length -eq 0) {
```
Comment (justincormack) [inline scripts/make.ps1]: I think you can use `$env:DOCKER_GITCOMMIT -eq $null` it is more idiomatic for powershell

**Comment by justincormack** [comment]
Needs a rebase

**Comment by dnephin** [comment]
Rebased, and fixed most of the comments (didn't touch the powershell)

**Comment by vdemeester** [comment]
@justincormack @dnephin rebased and updated the powershell :angel:

File: appveyor.yml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+clone_folder: c:\gopath\src\github.com\docker\cli
+
+environment:
+  GOPATH: c:\gopath
+  GOVERSION: 1.10
+  DEPVERSION: v0.4.1
+
+install:
+  - rmdir c:\go /s /q
+  - appveyor DownloadFile https://storage.googleapis.com/golang/go%GOVERSION%.windows-amd64.msi
```
Comment (mat007) [inline appveyor.yml]: Shouldn't that be prefixed with ```appveyor-retry``` just to be on the safe side ?

Comment (vdemeester) [reply appveyor.yml]: @mat007 noted, thanks :stuck_out_tongue: We never had any problem so far with that, so I'm fine merging like this. But if somehow it turns out we need to be more on the safe side, we'll use `appveyor-retry` then :wink:

**Pull request merged**